### PR TITLE
auth service

### LINKED ISSUE #277: Bootstrapping auth lib
This issue tracks the discussion and the first implementation around the auth lib and guard (see tracking issue or helpwave/web for an outline).

The auth logic around fake-tokens and OIDC redirects/callbacks should be handled in a library with a well-defined API. This API gets consumed by our apps (helpwave tasks). In helpwave/web, every time a caller want to get the user context, the session data gets checked for validity (outdated, invalid token). If unsuccessful, the user starts the sign-in process either with a Fake-Token or via our OIDC provider.

Could we implement such a structure in a lib for mobile? First, we would focus on the Fake-Tokens.

### PR DESCRIPTION
**Pull request by DasProffi**
## Which issues does this pull request close?
closes #277

### CODE REVIEW

File: packages/helpwave_service/lib/src/auth/authentication_service.dart [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+  factory AuthenticationService() => _authenticationService;
+
+  urlLauncher(Uri url) async {
+    await launchUrl(url, mode: LaunchMode.inAppWebView);
+  }
+
+  authenticate(
+    String discoveryUrl,
+    String redirectUrl, {
```
Comment (MaxSchaefer) [inline packages/helpwave_service/lib/src/auth/authentication_service.dart:17]: The redirectUrl should be non-configurable by the caller due to the fact that the caller should not create its own callback server.

File: packages/helpwave_service/lib/src/auth/authentication_service.dart [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  AuthenticationService._internal();
+
+  factory AuthenticationService() => _authenticationService;
+
+  urlLauncher(Uri url) async {
+    await launchUrl(url, mode: LaunchMode.inAppWebView);
+  }
+
+  authenticate(
+    String discoveryUrl,
```
Comment (MaxSchaefer) [inline packages/helpwave_service/lib/src/auth/authentication_service.dart:16]: We could fallback here as well to the default

File: packages/helpwave_service/lib/src/auth/authentication_service.dart [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    );
+
+    Credential? c;
+    // starts the authentication
+    try {
+      c = await authenticator.authorize();
+    } catch (e) {
+      print(e);
+    } finally {
+      // TODO this does't work :(
```
Comment (MaxSchaefer) [inline packages/helpwave_service/lib/src/auth/authentication_service.dart:45]: Only on Android. See iOS via BrowserStack

File: apps/tasks/lib/screens/landing_screen.dart [hunk 4]
```diff
@@ -29,9 +29,7 @@ class LandingScreen extends StatelessWidget {
             context.localization!.loginSlogan,
             style: Theme.of(context).textTheme.labelLarge,
           ),
-          onPressed: () => Navigator.of(context).pushReplacement(
-            MaterialPageRoute(builder: (context) => const MyTasksScreen()),
-          ),
+          onPressed: () => {AuthenticationService().authenticate("https://auth.helpwave.de", "http://localhost:3000/")}//Navigator.of(context).pushReplacement(MaterialPageRoute(builder: (context) => const MyTasksScreen()),),
```
Comment (MaxSchaefer) [inline apps/tasks/lib/screens/landing_screen.dart:32]: Please add a second button instead until the fake-tokens are implemented

File: packages/helpwave_service/lib/src/auth/authentication_service.dart
Comment (MaxSchaefer) [inline packages/helpwave_service/lib/src/auth/authentication_service.dart:1]: `clientId` and `discoveryUrl` should be env/build variable

**Review by MaxSchaefer — CHANGES_REQUESTED** [review]

**Review by LauraLobster — APPROVED** [review]

File: packages/helpwave_service/lib/src/auth/authentication_service.dart
Comment (MaxSchaefer) [reply packages/helpwave_service/lib/src/auth/authentication_service.dart:1]: Add comment

File: packages/helpwave_service/lib/src/auth/authentication_service.dart [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+    UserInfo userInfo = await credential.getUserInfo();
+
+    return Identity(
+      credential: credential,
+      id: userInfo.subject,
+      email: userInfo.email ?? "",
+      name: userInfo.name ?? "",
+      nickName: userInfo.nickname ?? "",
+      organizations: /* TODO userInfo.getTyped("organizations") ?? */ [],
```
Comment (MaxSchaefer) [inline packages/helpwave_service/lib/src/auth/authentication_service.dart:72]: All of this should be non-empty

**Review by MaxSchaefer — APPROVED** [review]

File: packages/helpwave_service/lib/src/auth/authentication_service.dart
Comment (DasProffi) [reply packages/helpwave_service/lib/src/auth/authentication_service.dart:1]: See https://github.com/helpwave/mobile-app/issues/424

**Pull request merged**